pkg/storage/unified/resource/server.go

```diff
@@ -0,0 +1,537 @@
+package resource
+
+import (
+	context "context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/bwmarrin/snowflake"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+)
+
+// HACK!!! since requester is not behaving as expected....
+// we are not getting the same names right now
+const CHECK_USER_MATCH = false
+
+// Package-level errors.
+var (
+	ErrNotFound                  = errors.New("entity not found")
+	ErrOptimisticLockingFailed   = errors.New("optimistic locking failed")
+	ErrUserNotFoundInContext     = errors.New("user not found in context")
+	ErrUnableToReadResourceJSON  = errors.New("unable to read resource json")
+	ErrNextPageTokenNotSupported = errors.New("nextPageToken not yet supported")
+	ErrLimitNotSupported         = errors.New("limit not yet supported")
+	ErrNotImplementedYet         = errors.New("not implemented yet")
+)
+
+// ResourceServer implements all services
+type ResourceServer interface {
+	ResourceStoreServer
+	ResourceSearchServer
+	DiagnosticsServer
+	LifecycleHooks
+}
+
+type AppendingStore interface {
```

Comment:
2 things:
- Add support for Batch create/update/delete?
- Could this be storage.Interface?


On Batch create/update/delete: I was digging into how the go-client fake works, and it's based on [ObjectTracker](https://pkg.go.dev/k8s.io/client-go/testing#ObjectTracker) which is very similar to the interface we are building. If you look at Add, it allows passing a list to add items separately.

Could this be [storage.Interface](https://pkg.go.dev/k8s.io/apiserver/pkg/storage#Interface) ? This follows my previous comment. The storage API represent common fields as metadata (message/updateBy/...). If so, the underlying store should only be concerned about writing/versinoning raw data ? This looks a lot like storage.Interface.
Some thoughts:
- We would get working etcd implementation (see [New](https://github.com/kubernetes/apiserver/blob/v0.30.2/pkg/storage/etcd3/store.go#L99) converting an ETCD client into a storage.Interface). One we can use for tests.
- storing List of items are supported
- Forces us to be compatible with k8s apiserver at the storage layer.
-

Reply:
I am looking at the [storage.go](https://github.com/grafana/grafana/blob/549ecfe053ec3f2e72542037b605cd89e02fd0cd/pkg/storage/unified/apistore/storage.go#L56) converting the ResourceStoreClient into an storage.Interface. Should our ResourceStore proto also be a strict implementation of storage.Interface to avoid too much conversion logic in the backend ?

Reply:
See https://github.com/grafana/grafana/pull/89331  (note that merges into this PR)